File: ts-playwright/src/products/booker/steps.ts
// Layer 2 — restful-booker business steps (workflows a test cares about).
import { BookerApiClient } from './BookerApiClient';

export class BookerSteps {
  constructor(private client: BookerApiClient) {}

  async createPaidBooking(firstname: string, lastname: string, total = 150): Promise<number> {
    const resp = await this.client.create({
      firstname,
      lastname,
      totalprice: total,
      depositpaid: true,
      bookingdates: { checkin: '2025-01-01', checkout: '2025-01-05' },
      additionalneeds: 'Breakfast',
    });
    return (await resp.json()).bookingid as number;
  }

  async cancelBooking(id: number, admin: { username: string; password: string }): Promise<void> {
    // Cancelling needs auth — the client handles token injection for us.
    await this.client.auth(admin.username, admin.password);
    await this.client.delete(id);
  }
}

File: python-playwright/framework/products/booker/steps.py
"""Layer 2 — restful-booker business steps.

Workflows a test cares about ("create a paid booking", "cancel a booking"),
composed from the BookerApiClient. Tests call these and assert outcomes.
"""

from __future__ import annotations

from .client import BookerApiClient


class BookerSteps:
    def __init__(self, client: BookerApiClient) -> None:
        self.client = client

    def create_paid_booking(self, firstname: str, lastname: str, *, total: int = 150) -> int:
        booking = {
            "firstname": firstname,
            "lastname": lastname,
            "totalprice": total,
            "depositpaid": True,
            "bookingdates": {"checkin": "2025-01-01", "checkout": "2025-01-05"},
            "additionalneeds": "Breakfast",
        }
        resp = self.client.create(booking)
        return resp.json()["bookingid"]

    def cancel_booking(self, booking_id: int, admin: dict[str, str]) -> None:
        # Cancelling needs auth — the client handles token injection for us.
        self.client.auth(**admin)
        self.client.delete(booking_id)
